astrolabe/spec_runner.py

```diff
@@ -213,33 +210,40 @@ def run(self, persist_cluster=False):
         # Step-5: interrupt driver workload and capture streams
         LOGGER.info("Stopping workload executor [PID: {}]".format(
             worker_subprocess.pid))
-        os.kill(worker_subprocess.pid, self.sigint)
-        stdout, stderr = worker_subprocess.communicate()
-        LOGGER.info("Stopped workload executor")
+        os.killpg(worker_subprocess.pid, self.sigint)
+        stdout, stderr = worker_subprocess.communicate(timeout=10)
+        LOGGER.info("Stopped workload executor [exit code: {}]".format(
+            worker_subprocess.returncode))
 
         # Stop the timer
         timer.stop()
 
         # Step-6: compute xunit entry.
         junit_test = junitparser.TestCase(self.id)
         junit_test.time = timer.elapsed
-        if worker_subprocess.returncode != 0:
+
+        try:
+            err_info = json.loads(stderr)
+        except json.JSONDecodeError:
+            err_info = {'numErrors': '-1', 'numFailures': '-1'}
+
+        if int(err_info['numErrors']) or int(err_info['numFailures']) \
+                or worker_subprocess.returncode != 0:
             LOGGER.info("FAILED: {!r}".format(self.id))
-            self.failed = True
+            # Write xunit logs for failed tests.
             errmsg = """
             Number of errors: {numErrors}
-            Number of failures: {numFailures}    
-            """
-            try:
-                err_info = json.loads(stderr)
-                junit_test.result = junitparser.Failure(
-                    errmsg.format(**err_info))
-            except json.JSONDecodeError:
-                junit_test.result = junitparser.Error(encode_cdata(stderr))
+            Number of failures: {numFailures}
```

Comment:
This error message uses multi-line string so it includes all the whitespace. Do we want it to be indented this much?

Reply:
Good point, I will remove the whitespace.